File: LeetCode/AddTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(None)
        node = head
        excess = 0
        while l1 and l2:
            val = excess + l1.val + l2.val
            excess, val = val // 10, val % 10
            node.next = ListNode(val)
            l1, l2 = l1.next, l2.next
            node = node.next
        else:
            n = None
            if l1:
                n = l1
            elif l2:
                n = l2
            if excess:
                node.next = Solution.addTwoNumbers(Solution, ListNode(excess), n)
            else:
                node.next = n
        return head.next

File: Java/LeetCode/Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) { val = x; }
 * }
 */
package LeetCode;

class Solution_Add_Two_Numbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return partialAdd(l1, l2, 0);
    }

    public ListNode partialAdd(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null) return carry == 0 ? null : new ListNode(carry);
        int val1 = l1 == null ? 0 : l1.val;
        int val2 = l2 == null ? 0 : l2.val;
        int node_val = (val1 + val2 + carry) % 10;
        int next_carry = (val1 + val2 + carry) / 10;
        ListNode node = new ListNode(node_val);
        node.next = partialAdd(l1 == null ? null : l1.next, l2 == null ? null : l2.next, next_carry);
        return node;
    }
}